code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -185,3 +185,127 @@ pub fn compute_deposit_lp(
+}
+
+/// Paramaters:
+/// * `lp_token_details` - Vec of tuples containing `(token_deposit, token_balance, token_weight)`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 6
+fn compute_first_deposit_lp_<T: PerThing>(
+	lp_token_details: Vec<(u128, u128, T)>,
+	_f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	let k: u128 = lp_token_details.len().try_into().map_err(|_| ArithmeticError::Overflow)?;
+	let product = lp_token_details
+		.iter()
+		.try_fold(1, |product, (_d_i, b_i, _w_i)| product.safe_mul(b_i))?;
+
+	// REVIEW: How is fee applied here?
+	Ok((k.safe_mul(&product)?, 0))
+}
+
+/// Paramaters:
+/// * `p_supply` -
+/// * `d_k` - Deposit of token `k`
+/// * `b_k` - Balance of token `k`
+/// * `w_k` - Weight of token `k`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 5
+fn compute_existing_deposit_lp_<T: PerThing>(
+	p_supply: u128,
+	d_k: u128,
+	b_k: u128,
+	w_k: T,
+	f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	let p_supply = Decimal::from_u128(p_supply).ok_or(ArithmeticError::Overflow)?;
+	let d_k = Decimal::from_u128(d_k).ok_or(ArithmeticError::Overflow)?;
+	let b_k = Decimal::from_u128(b_k).ok_or(ArithmeticError::Overflow)?;
+	let w_k = decimal_from_per_thing(w_k)?;
+
+	let left_from_fee =
+		if f.is_zero() { Decimal::ONE } else { decimal_from_per_thing(f.left_from_one())? };
+	let d_k_left_from_fee = d_k.safe_mul(&left_from_fee)?;
+
+	let base = d_k_left_from_fee.safe_add(&b_k)?.safe_div(&b_k)?;
+	let power = base.checked_powd(w_k).ok_or(ArithmeticError::Overflow)?;
+	let ratio = power.safe_sub(&Decimal::ONE)?;
+
+	let issued = p_supply.safe_mul(&ratio)?.to_u128().ok_or(ArithmeticError::Overflow)?;
+	let fee = d_k.safe_sub(&d_k_left_from_fee)?.to_u128().ok_or(ArithmeticError::Overflow)?;
+
+	Ok((issued, fee))
+}
+
+/// Compute lp to mint for a pool.
+///
+/// If `Ok`, returns a tuple containing (lp_to_mind, fee).
+/// `lp_token_details` must not be empty.
+///
+/// Paramaters:
+/// * `p_supply` - Existing supply of LPT tokens
+/// * `lp_token_details` - Vec of tuples containing `(token_deposit, token_balance, token_weight)`
+/// * `f` - Fee
+pub fn compute_deposit_lp_new<T: PerThing>(
+	p_supply: u128,
+	lp_token_details: Vec<(u128, u128, T)>,
+	f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	ensure!(!lp_token_details.is_empty(), ConstantProductAmmError::InvalidTokensList);
+
+	let is_first_deposit = p_supply.is_zero();
+
+	if is_first_deposit {
+		compute_first_deposit_lp_(lp_token_details, f)
+	} else {
+		// REVIEW: Should this be done at the scale of `lp_token_details.len()`?
+		let token_k = lp_token_details.last().expect("List is not empty; QED");
+		compute_existing_deposit_lp_(p_supply, token_k.0, token_k.1, token_k.2, f)
+	}
+}
```

Comment:
> Should this be done at the scale of `lp_token_details.len()`?

Should the calculation for existing deposits be done at the scale of the input array? The current math in the docs does not infer it but does seem like something to consider.

Reply:
We need to do some optimization here as `lp_token_details` being an array can be confusing for a caller. We do not expect this function to be called with an array of token details for the general case, it will always be a single asset deposit(`d_k`), balance(`b_k`) and the weight (`w_k`). This is we only support either adding a liquidity in a single asset (`b_k, d_k, w_k` for a single asset) or adding with all the assets in weight proportion (any on of the `(b_k, d_k)` with `w_k = 1`). Therefore I'm thinking we should remove the `first_deposit` decision here and make the caller make that decision and call the correct function. We can then also get rid of `compute_deposit_lp_new` I think.